modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,708 @@
+
+      // Group role for Canvas users who are an instructor in one more courses
+      if (isInstructor)
+        roles.add(new JaxbRole(Group.ROLE_PREFIX + "CANVAS_INSTRUCTOR", jaxbOrganization, "Canvas Instructors",
+            Role.Type.EXTERNAL_GROUP));
+
+      logger.debug("Returning JaxbRoles: " + roles);
+
+      // JaxbUser(String userId, String password, String name, String email, String provider, boolean canLogin,
+      // JaxbOrganization organization, Set<JaxbRole> roles)
+      User user = new JaxbUser(userId, null, displayName, email, PROVIDER_NAME, jaxbOrganization, roles);
+
+      cache.put(userId, user);
+      logger.debug("Returning user {}", userId);
+
+      return user;
+
+    } finally {
+      currentThread.setContextClassLoader(originalClassloader);
+    }
+
+  }
+
+  /**
+   * Verify that the user exists
+   * Query with /api/v1/users/sis_user_id:{userId}
+   *
+   * @param userId Canvas User ID
+   */
+  private boolean verifyCanvasUser(String userId) {
+    logger.debug("verifyCanvasUser({})", userId);
+    try {
+      if ((userPattern != null) && !userId.matches(userPattern)) {
+        logger.debug("Verify user {} failed regexp {}", userId, userPattern);
+        return false;
+      }
+    } catch (PatternSyntaxException e) {
+      logger.warn("Invalid regular expression for user pattern {} - disabling checks, exception: {}", userPattern, e);
+      userPattern = null;
+    }
+
+    int code;
+
+    String urlString = canvasUrl + "/api/v1/users/sis_user_id:" + userId;
+    try {
+      URL url = new URL(urlString);
+      logger.debug("Verifying user: {} using API", url);
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestMethod("GET");
+      connection.setRequestProperty("Authorization", "Bearer " + canvasToken);
+      connection.setRequestProperty("User-Agent", OC_USERAGENT);
+
+      connection.connect();
+      code = connection.getResponseCode();
+    } catch (Exception e) {
+      logger.warn("Exception verifying Canvas user {} at {}: {}", userId, urlString, e);
+      return false;
+    }
+    // HTTP OK 200 for course exists, return false for everything else (typically 404 not found)
+    return (code == 200);
+  }
+
+  /**
+   * Verify that the course exists
+   * Query with /api/v1/courses/{courseId}
+   *
+   * @param courseId Canvas Course ID
+   */
+  private boolean verifyCanvasCourse(String courseId) {
+    // We could additionally cache positive and negative courseId lookup results here
+    logger.debug("verifyCanvasCourse({})", courseId);
+    try {
+      if ((coursePattern != null) && !courseId.matches(coursePattern)) {
+        logger.debug("Verify course {} failed regexp {}", courseId, coursePattern);
+        return false;
+      }
+    } catch (PatternSyntaxException e) {
+      logger.warn("Invalid regular expression for course pattern {} - disabling checks, exception: {}", coursePattern
```

Comment:
```suggestion
      logger.warn("Invalid regular expression for course pattern {} - disabling checks", coursePattern
```

Same question here: Do we really want to have a fallback?